### PR TITLE
Serialization changes to reduce library size

### PR DESCRIPTION
**Pull request by benjaminulmer**
The grid-based solution selection we're working to support requires orders of magnitude more sizes in the library than we normally have in rocBLAS (~100k vs 1-10k). This PR makes some changes to the serialization so this increase in sizes won't be as impactful on the final file size.

This PR has three changes:

**1) Removing code-gen parameters from solutions**
The only code-gem parameter that was used was the library logic index, so this has been retained as its own dedicated field.

**2) Making SingleSolutionLibrary just an index**
This changes the serialization for the solution indices from `{value: {Type: Single, Value: <idx>}}` to just `{index: <idx>}`. I'm not totally happy with the current implementation of this, though, and would appreciate feedback with ideas or suggestions.

**3) Remove the speed for GridBased libraries**
GridBased libraries do not need the speed entry for breaking ties, so it has been removed for just that library type

### CODE REVIEW

File: Tensile/Source/lib/include/Tensile/Serialization/MatchingLibrary.hpp [hunk 1]
```diff
@@ -217,9 +218,43 @@ namespace Tensile
...
             {
+                int32_t index = -1;
                 iot::mapRequired(io, "key", entry.key);
-                iot::mapRequired(io, "value", entry.value);
-                iot::mapRequired(io, "speed", entry.speed);
+                iot::mapOptional(io, "speed", entry.speed);
+                iot::mapOptional(io, "index", index);
+
+                // TODO there is probably a way to make this properly general
+                if(index != -1)
```
[the remark this answers is not in the document]
Comment (nielenventer) [reply Tensile/Source/lib/include/Tensile/Serialization/MatchingLibrary.hpp:227]: Ya seems like if there are different expected params they could also be different library types? Like a GridMatchingLibrary?

[the remark this answers is not in the document]
Comment (benjaminulmer) [reply Tensile/Source/lib/include/Tensile/Serialization/MatchingLibrary.hpp:227]: The else block should handle cases where we still want a fully general library entry. The lack of generality is that this is hardcoded on ContractionProblem and ContractionSolution (vs template parameters). I couldn't figure out how to navigate the template web well enough to make it work as template parameters.

[the remark this answers is not in the document]
Comment (nielenventer) [reply Tensile/Source/lib/include/Tensile/Serialization/MatchingLibrary.hpp:227]: Ok, ya it's a bit clunky but not the end of the world. I think it reflects the fact that we're transitioning/testing this new approach and hopefully the natural conclusion will be to pick one and then code will clean up.

**Review by nielenventer — APPROVED** [review]

**Review by JeremyAdamHart — APPROVED** [review]
The change in .dat file size looks really good, I'm seeing about 1/3rd the file size. I'm not seeing any red flags in the code changes so LGTM

**Pull request merged**